Wizard: blog post, docs and agent-based onboarding (#11318)
PostHog/posthog.com: contents/blog/envoy-wizard-llm-agent.mdx

See, **these agents can run terminal commands**, so any CLI-based program can adopt this pattern.

With this deterministic surface introduced to the agent workflow, you can reliably and predictably:

- Enforce progression conditions, not moving to another step until requirements are satisfied  
- Enforce correct versioning
- Use API interactions to populate account secrets and other details, while enforcing good version control hygiene  
- Introduce specific fragments of code verbatim
- Inject up-to-date documentation and agent rules into the project to course-correct future agent sessions  
- Do your own LLM-based transformations on code, using prompts, context and model versions that are known to be reliable for your goals
- Enjoy same-day updates to agent workflows, just like the rest of your software

Think of it like a sheepdog that keeps the agent session within a territory that is productive rather than stupid, adhering to best practices **you** as a developer can describe in code.

Naturally, we're building our own envoy already. It's a beta, but it's already very cool and the feedback has been promising.

## Example code: the PostHog wizard

> The PostHog [wizard](https://github.com/PostHog/wizard/tree/main) is an interactive command line tool that automates the basics of a PostHog integration. It grabs an API key from a user’s account and uses it to populate a .env file.

Reviewer: We can replace the link with just `https://github.com/PostHog/wizard`

From here it uses LLMs and consistent prompting to read through the user’s code and integrate PostHog correctly.

It turns out this is handy behavior: other LLM workflows were hallucinating PostHog API keys and agents were using out-of-date code to perform integrations.

To make the wizard into an effective **envoy**, all we had to do was add an argument that suppressed the interactivity, following the default path all the way through the workflow. The agent just runs the command in its terminal.

From here, anyone using an AI agent can copy and paste one prompt and get a working integration in about 90 seconds. It’s the best of both worlds:

- No hallucinations of API keys  
- No out of date patterns  

> - Full vibe coding cruise control

Reviewer: 💖

Author: womp womp, that didn't take long. I can't access the link, DM it to me?

Maybe it's "agent coding" then

---

> title: "Steal this AI code"

Author: we can do that!

---
title: "Please steal this AI code"
date: 2025-04-23
author:
  - danilo-campos
featuredImage: >-
  https://res.cloudinary.com/dmukukwp6/image/upload/steal_this_code_90451c15ca.png
featuredImageType: full
tags:
    - Engineering
---

import Prompt from '../docs/integrate/_snippets/agent-prompt.mdx'

AI agents are taking over the workflows of software development. They can burn through boilerplate and tedium fast, getting working prototypes up in minutes.

But LLM codegen has two annoying problems...

The first? **There are so many stupid ways to build software**.

When an agent sets out to write code, it has many viable paths to both solve the problem and introduce mistakes. Mistakes are especially common for newer, upstart projects.

Scale is one way to compensate: fill the training set with enough statistical weight that favors correct implementations, you’re less likely to see the boneheaded stuff.

> But that’s a lot of effort, plus significant lag time, and you want correct agent behavior *today*.

Author: Perhaps:
> But that’s a lot of effort, plus significant lag time as you await scraping and training. You want correct agent behavior *today*.

The second problem? **LLMs are so expensive**.

Training models is expensive, GPUs are expensive. Sometimes it feels like the best case scenario is a successful donation machine piped to Anthropic's coffers.

But: good news. You can join this revolution in a way that won't drain your bank account.

Build an **envoy**: perfectly conventional software *you control* that developers can inject into their agent sessions with just a prompt.

## What's an envoy?

> To coin a term, an **envoy** is code that meets an AI agent on your behalf.
> 
> See, **these agents can run terminal commands**, so any CLI-based program can adopt this pattern. Envoys provide a side entrance into any agent session. A CLI app can print things to the terminal, adding to the context the agent has available. This is in addition, of course, to all the modifications such a tool can make to the project's code and configuration.

Reviewer 2: I don't fully understand what "side entrance" means or why it being a CLI is that big of a deal, I feel like this gets to the core of what an envoy is more:

```suggestion
To coin a term, an **envoy** is code that meets an AI agent on your behalf.

See, many agents can now run terminal commands. This means they can install packages and libraries when appropriate. This also means they can cause havoc installing packages and libraries poorly.

An envoy helps prevent this by guiding agents into doing this well. It works by printing things to the terminal which adds to the context the agent has available. This is in addition, of course, to all the modifications such a tool can make to the project's code and configuration.
```

Reviewer 2: This is really the core of the article and will make or break whether "envoy" becomes a term or not, so worth getting it right :)

Author: > See, many agents can now run terminal commands. This means they can install packages and libraries when appropriate. This also means they can cause havoc installing packages and libraries poorly.

This isn't the causal link, though. They could (and do) make these errors even if they can't run terminal commands.

Reviewer 2: What is a causal link?

Author: > This also means they can cause havoc installing packages and libraries poorly.

there is nothing about being able to run terminal commands that causes the above ^

With this deterministic surface introduced to the agent workflow, you can reliably and predictably:

- Enforce progression conditions, not moving to another step until requirements are satisfied  
- Enforce correct versioning

> - Use API interactions to populate account secrets and other details, while enforcing good version control hygiene  

Reviewer 2: You already mention versioning above
```suggestion
- Use API interactions to populate account secrets and other details  
```

Author: Not versioning, version control hygiene, to wit: putting secrets into files that are .gitignore'd

- Introduce specific fragments of code verbatim
- Inject up-to-date documentation and agent rules into the project to course-correct future agent sessions  
- Configure MCPs and other project-specific settings
- Do your own LLM-based transformations on code, using prompts, context and model versions that you carefully select and design for reliable outcomes
- Enjoy same-day updates to agent workflows, just like the rest of your software

> - Provide up-to-date information to the agent specific to the project

Reviewer 2: These two are sort of the same point... Combine into one?

Author: They are distinct, but I've updated to make that clearer

> ![Envoy diagram](https://res.cloudinary.com/dmukukwp6/image/upload/envoy_diagram_9dd36b6c32.png)

Reviewer 2: Maybe adding more context on what the guardrails are doing?

> ![Envoy diagram](https://res.cloudinary.com/dmukukwp6/image/upload/envoy_diagram_9dd36b6c32.png)

Reviewer 2: I definitely think we should have a graphic like this. It can be this simple, just should be in PostHog style.

> ![Envoy diagram](https://res.cloudinary.com/dmukukwp6/image/upload/envoy_diagram_9dd36b6c32.png)

Author: How do you think that would look? I actively, deliberately don't want the text set in Squeak, for example. It's too… jokey, non-serious for a technical post about a design pattern whose adoption will take some serious time and effort, if that makes any sense.

Think of it like a sheepdog that keeps the agent session within a territory that is productive rather than stupid, adhering to best practices **you** as a developer can describe in code.

Naturally, we're building our own envoy already. It's a beta, but it's already very cool and the feedback is encouraging.

## Example code: the PostHog wizard

> The PostHog [wizard](https://github.com/PostHog/wizard) started its life as an interactive command line tool for automating the basics of a PostHog integration. It grabs an API key from a user’s account and uses it to populate a .env file.

Reviewer 2: ```suggestion
The PostHog [wizard](https://github.com/PostHog/wizard) started its life as an interactive command line tool for automating the basics of a PostHog integration. It grabs an API key from a user’s account and uses it to populate a `.env` file.
```

It turns out this is handy behavior: other LLM workflows were *hallucinating* PostHog API keys and agents were using out-of-date code to perform integrations.

To make the wizard into an effective **envoy**, all we had to do was add an argument that suppressed the interactivity, following the default path all the way through the workflow. The agent just runs the command in its terminal, on the user's behalf.

![Agent session](https://res.cloudinary.com/dmukukwp6/image/upload/agent_session_04407be548.gif)

Now anyone using an AI agent can copy and paste one prompt and get a working integration in about 90 seconds. It’s the best of all worlds:

- No hallucinations of API keys  
- No out of date patterns  
- Smooth, automated code generation

We can ship as many updates to the wizard as fast as we like, and the improved behavior will be available to developers who use agents instantly, with no other configuration.

Try it yourself. Paste this into the agent chat for Cursor, Bolt, and others:

<Prompt />

We’ve got support for Next.js, React, Svelte, and React Native, with others coming soon.

## Steal this code

> Our [Joshua Snyder](https://posthog.com/community/profiles/32497) has been working hard to increase the surface of the wizard and make it more reliable. The project code is tidy, and it even demonstrates how to integrate product analytics into a node-based CLI tool.

Reviewer 2: ```suggestion
[Joshua](https://posthog.com/community/profiles/32497) has been working hard to increase the surface of the wizard and make it more reliable. The project code is tidy, and it even demonstrates how to integrate product analytics into a Node-based CLI tool.
```

> We get reporting on completed wizard runs alongside all the rest of our data. You can learn a lot from this project!

Reviewer 2: I don't understand this point or how it is valuable to readers.

Author: It's essentially:

"Did you know you can add analytics to your CLI surfaces? you can, here's what you get for it, check out the implementation to learn the details"

I think it's the most exciting part of this tbh, is there some other way to draw that out?

Reviewer 2: Ahh, maybe this?
```suggestion
The project also shows how you can implement PostHog in your CLI surfaces, which gets us all sort of useful data, like which libraries and frameworks are most popular. You can do this too!
```
